ejector/ejection_manager.go

```diff
@@ -0,0 +1,408 @@
+		startEjectionThrottleFull:            config.StartEjectionThrottleFull,
+	}
+
+	for _, addr := range config.DoNotEjectTheseValidators {
+		em.ejectionBlacklist[geth.HexToAddress(addr)] = struct{}{}
+	}
+
+	// Set up a throttle for quorum 0. We will always have a quorum 0, and this allows us to check to see
+	// if the throttle config is valid. Checking here lets us assume it is valid later on.
+	var err error
+	em.quorumRateLimits[0], err = ratelimit.NewLeakyBucket(
+		config.EjectionThrottle,
+		config.EjectionThrottleTimePeriod,
+		config.StartEjectionThrottleFull,
+		ratelimit.OverfillOncePermitted,
+		timeSource())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create leaky bucket: %w", err)
+	}
+
+	return em, nil
+}
+
+func (em *ejectionManager) BeginEjection(
+	validatorAddress geth.Address,
+	stakes map[core.QuorumID]float64,
+) {
+
+	// Check to see if the validator is blacklisted.
+	if _, blacklisted := em.ejectionBlacklist[validatorAddress]; blacklisted {
+		em.logger.Infof("validator %s is blacklisted from ejection, will not begin ejection",
+			validatorAddress.Hex())
+		return
+	}
+
+	// Check to see if we are already in the process of ejecting this validator.
+	if _, inProgress := em.ejectionsInProgress[validatorAddress]; inProgress {
+		em.logger.Infof("ejection already in progress for validator %s, will not begin ejection",
+			validatorAddress.Hex())
+		return
+	}
+
+	// Check to see if we have recently attempted to eject this validator.
+	if _, recentlyEjected := em.recentEjectionTimes[validatorAddress]; recentlyEjected {
+		em.logger.Infof("recent ejection attempt for validator %s, will not begin ejection",
+			validatorAddress.Hex())
+		return
+	}
+
+	// Check to see if there is already an ejection in progress on-chain for this validator.
+	inProgress, err := em.transactor.IsEjectionInProgress(em.ctx, validatorAddress)
+	if err != nil {
+		em.logger.Errorf("failed to check ejection status for validator %s, will not begin ejection: %v",
+			validatorAddress.Hex(), err)
+		return
+	}
+
+	// Check if we are prevented from starting an ejection by rate limiting.
+	allowedByRateLimits := em.checkRateLimits(validatorAddress, stakes)
+	if !allowedByRateLimits {
+		// Rate limiting prevents us from starting an ejection at this time.
+		// checkRateLimits() will have logged the reason, since it has more context.
+		return
+	}
+
+	if inProgress {
+		// An ejection is already in progress. Record it, and we can try to finalize it later.
+		em.logger.Infof("ejection already in progress on-chain for validator %s, "+
+			"will not begin ejection but will attempt to finalize",
+			validatorAddress.Hex())
+	} else {
+		// Start a new ejection.
+		err = em.transactor.StartEjection(em.ctx, validatorAddress)
+		if err != nil {
+			em.logger.Errorf("failed to start ejection for validator %s: %v", validatorAddress.Hex(), err)
+			em.cleanUpFailedEjection(validatorAddress, stakes)
```

Comment:
What if an error is thrown in `StartEjection` before the debit actually completes? `cleanUpFailedEjection` would then refund a debit that never happened